pkg/controller/certificatesigningrequests/selfsigned/selfsigned.go

```diff
@@ -0,0 +1,206 @@
+// CertificateSigningRequests must have the
+// "experimental.cert-manager.io/private-key-secret-name" annotation present to
+// be signed. This annotation must reference a valid Secret containing a CA
+// private key for signing.
+func (s *SelfSigned) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+
+	secretName, ok := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestPrivateKeyAnnotationKey]
+	if !ok || len(secretName) == 0 {
+		message := fmt.Sprintf("Missing private key reference annotation: %q", experimentalapi.CertificateSigningRequestPrivateKeyAnnotationKey)
+		log.Error(errors.New(message), "")
+		s.recorder.Event(csr, corev1.EventTypeWarning, "MissingAnnotation", message)
+		util.CertificateSigningRequestSetFailed(csr, "MissingAnnotation", message)
+		_, err := s.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	resourceNamespace := s.issuerOptions.ResourceNamespace(issuerObj)
+
+	privatekey, err := kube.SecretTLSKey(ctx, s.secretsLister, resourceNamespace, secretName)
+	if apierrors.IsNotFound(err) {
+		message := fmt.Sprintf("Referenced Secret %s/%s not found", resourceNamespace, secretName)
+		log.Error(err, message)
+		s.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		util.CertificateSigningRequestSetFailed(csr, "SecretNotFound", message)
+		_, err = s.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	if cmerrors.IsInvalidData(err) {
+		message := fmt.Sprintf("Failed to parse signing CA key from secret %s/%s", resourceNamespace, secretName)
+		log.Error(err, message)
+		s.recorder.Eventf(csr, corev1.EventTypeWarning, "ErrorParsingKey", "%s: %s", message, err)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorParsingKey", message)
+		_, err = s.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	if err != nil {
+		// We are probably in a network error here so we should backoff and retry
+		message := fmt.Sprintf("Failed to get certificate CA key from secret %s/%s", resourceNamespace, secretName)
+		log.Error(err, message)
+		s.recorder.Eventf(csr, corev1.EventTypeWarning, "ErrorGettingSecret", "%s: %s", message, err)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorGettingSecret", message)
+		_, err = s.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	template, err := pki.GenerateTemplateFromCertificateSigningRequest(csr)
+	if err != nil {
+		message := fmt.Sprintf("Error generating certificate template: %s", err)
+		log.Error(err, message)
+		s.recorder.Event(csr, corev1.EventTypeWarning, "ErrorGenerating", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorGenerating", message)
+		_, err = s.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	template.CRLDistributionPoints = issuerObj.GetSpec().SelfSigned.CRLDistributionPoints
```

Comment:
suggestion: I firmly believe that adding CRL distribution points was a mistake for the SelfSigned issuer; they don't make any sense to me for SelfSigned issuers. Obviously if we need 100% parity with `CertificateRequest`s then we have to add this line, but this seems a really good opportunity to get started with removing them.

(if we can't remove them and we _have_ to keep them in, then we should probably add tests for them - currently, I don't think they're tested in this PR)

```suggestion
	// NB: CSRs don't support CRLDistributionPoints for SelfSigned issuers.
	// Their inclusion for CertificateRequests was a mistake, and the field is ignored for CSR resources.
```

Reply:
Indeed we should keep it in to have feature parity. This is tested here https://github.com/jetstack/cert-manager/pull/4100/files#diff-cea0019cb83d2041682852bbebe3dbdf7c8482e941fcd9c64b81b8e6241bb7beR758

Reply:
ah my bad, I totally missed that test.

> Indeed we should keep it in to have feature parity.

I think my point is it's not a feature, it's just a mistake. but yeah, if it's gotta be there it's gotta be there.

Reply:
Personally I don't think it is a mistake. The original issue specifically wanted them available for SelfSigned https://github.com/jetstack/cert-manager/issues/2612

Whether that brings any "real value" is another thing, however some organisations seem to have policy that they must be present on all certificates regardless.

Reply:
i guess this isn't the place for a discussion about why it's a mistake